IdentityCore/src/requests/MSIDSilentTokenRequest.m

```diff
@@ -235,126 +217,132 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
-                 }
-             }];
+    completionHandler(nil, MSIDInvalidRefreshTokenType, nil);
 }
 
-- (void)tryAppRefreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)multiResourceRefreshToken
-           completionBlock:(nonnull MSIDRequestCompletionBlock)completionBlock
+- (void)tryRefreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)refreshToken
+              tokenType:(MSIDRefreshTokenTypes)tokenType
+        completionBlock:(nonnull MSIDRequestCompletionBlock)completionBlock
 {
-    if (!multiResourceRefreshToken)
-    {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Refresh token wasn't found, user interaction is required.");
-        
-        NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", nil, nil, nil, self.requestParameters.correlationId, nil, YES);
-        completionBlock(nil, interactionError);
-        return;
-    }
+    BOOL isAppRefreshToken = tokenType == MSIDAppRefreshTokenType;
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelVerbose, self.requestParameters, @"Trying to acquire access token using %@ for clientId %@, authority %@, account %@", (!isAppRefreshToken ? @"App Refresh Token" : @"Family Refresh Token"), self.requestParameters.authority, self.requestParameters.clientId, self.requestParameters.accountIdentifier.maskedHomeAccountId);
     
-    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Found Refresh token, using it...");
-    [self refreshAccessToken:multiResourceRefreshToken
-             completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
-                 if (error)
-                 {
-                     //Check if server returns invalid_grant or invalid_request
-                     if ([self isErrorRecoverableByUserInteraction:error])
-                     {
-                         NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", error.msidOauthError, error.msidSubError, error, self.requestParameters.correlationId, nil, YES);
-
-                         completionBlock(nil, interactionError);
-                         return;
-                     }
-                     else
-                     {
-                         completionBlock(nil, error);
-                     }
-                 }
-                 else
-                 {
-                     completionBlock(result, nil);
-                 }
-             }];
+    // When using ART or FRT, it will go through the same method below, and handle differently within the completion block
+    [self refreshToken:refreshToken
+       completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
+        if (error)
+        {
+            if ([self isErrorRecoverableByUserInteraction:error])
+            {
+                NSError *msidError = nil;
+                
+                if (isAppRefreshToken)
```

Comment:
This code is a bit hard to follow. Would it make sense to have different error handling for app refresh token and family refresh token?

Reply:
Refactored the error handling into two separated functions one for ART and the other is for FRT